File: LinkedListCycleTwo.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

# Using Two pointer approach (fast and slow pointers) to detect the cycle
# After detecting reset one of the pointer to determine the start of the cycle
# Time Complexity is O(n)
# Space Complexity is O(1)
class Solution(object):
    def detectCycle(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        if head is None:
            return None
        slow = head
        fast = head
        hasCycle = False
        
        while(fast and fast.next):
            slow = slow.next
            fast = fast.next.next
            if(slow == fast):
                hasCycle = True
                break
            
            
        
        if(not hasCycle):
            return None
        
        slow = head
        while(slow!=fast):
            slow = slow.next
            fast = fast.next
        return slow

File: linkedlist_cycle_two.java
//TC: O(n)
//SC: O(1)
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode detectCycle(ListNode head) {
        boolean flag=false;
        ListNode slow;
        ListNode fast;
        slow=head;
        fast=head;
        while(fast!=null&&fast.next!=null)
        {
        
            fast=fast.next.next;
            slow=slow.next;
                 if(fast==slow)
           {
               flag=true;
               break;
           }
         
        }
        if(!flag)
        {
            return null;
        }
        fast=head;
        while(fast!=slow)
        {
            fast=fast.next;
            slow=slow.next;
        }
        return fast;
        
    }
}